[GTK][MiniBrowser] Amazon Luna landing page doesn't render

cadubentzen opened the pull request:
#### d16890bc3dd257a76fbd6dc10e17bac2f95cc29b
<pre>
[GTK][MiniBrowser] Amazon Luna landing page doesn&apos;t render
<a href="https://bugs.webkit.org/show_bug.cgi?id=276585">https://bugs.webkit.org/show_bug.cgi?id=276585</a>

Reviewed by Michael Catanzaro.

Amazon Luna (luna.amazon.com) assumes that the user agent is on MacOS if
`window.webkit` is defined, leading to JS calls that only work in the
Apple Mac WebKit port.

`window.webkit` was defined in the GTK MiniBrowser because it registers
a script message handler &quot;aboutData&quot;, which makes
`window.webkit.messageHandlers.aboutData.postMessage(msg)` available to
scripts. However, this is only useful for the &quot;about:data&quot; custom page
that is defined in the MiniBrowser.

To have MiniBrowser still working like before but also supporting Amazon
Luna, only register the script message handler when loading &quot;about:*&quot; pages.

* Tools/MiniBrowser/gtk/BrowserTab.c:
(toggleAboutDataScriptMessageHandler):
(loadChanged):
* Tools/MiniBrowser/gtk/main.c:
(activate):

Canonical link: <a href="https://commits.webkit.org/281898@main">https://commits.webkit.org/281898@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/83cd728aacc7eabd6176f13684c397fa54c7e26c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61367 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40727 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13951 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65317 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11915 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63497 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48405 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12190 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/49570 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8271 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64156 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37861 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53153 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30413 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34522 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10381 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10828 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56330 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10681 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67049 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5313 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/10455 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56955 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5336 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53118 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57161 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13680 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4385 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36531 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37614 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38708 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37358 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Tools/MiniBrowser/gtk/BrowserTab.c:
```diff
@@ -146,6 +146,21 @@ static void loadChanged(WebKitWebView *webView, WebKitLoadEvent loadEvent, Brows
     if (loadEvent != WEBKIT_LOAD_STARTED)
         return;
 
+    WebKitUserContentManager *userContentManager = webkit_web_view_get_user_content_manager(webView);
```
This will run a few times for each top-level page load event except `WEBKIT_LOAD_STARTED` (skipped above). While this is not much of an issue (re-registering repeatedly should be idempotent) it is unneeded busy-work, I think it would be better to only do this during `WEBKIT_LOAD_COMMITTED` because at that point the URI of the page being loaded is already known to be definitive and will not change again for the current load.

cadubentzen replied:
> This will run a few times for each top-level page load event except WEBKIT_LOAD_STARTED (skipped above)

@aperezdc I'm not sure I follow. Isn't the code above _skipping every load-changed event except_ `WEBKIT_LOAD_STARTED`? So the code is only run for `WEBKIT_LOAD_STARTED`, which iiuc only runs once per top-level page, not the other way around :thinking:

aperezdc replied:
@cadubentzen Ah, you're right, sorry about the confusion. Nevertheless, it would be still better to use `_LOAD_COMMITTED` for checks on the URI, because it may change after `_LOAD_STARTED`, for example due to redirections.

cadubentzen replied:
Alright, makes sense then! Will do, thanks :)

aperezdc reviewed (changes requested):
Approach LGTM with a nit, please take a look.

mcatanzaro reviewed (approved):
Wow, this bug is pretty intense.

I agree this is a good solution though. The strategy used by Epiphany is that script message handlers are either registered in a non-default script world, or else registered temporarily only when displaying internal pages, to prevent web content from seeing them. Matching that behavior seems good.

I had never imagined it could be a web compatibility issue, though. Bonus points are available if you add some documentation `WebKitUserContentManager` to warn about this problem, so that app developers are less likely to mess up. (We even had one unstable Epiphany release where we exposed the internal password manager interface in the default script world, so websites could just enumerate everything stored in your password manager. Oops....)

cadubentzen committed:
[GTK][MiniBrowser] Amazon Luna landing page doesn't render
https://bugs.webkit.org/show_bug.cgi?id=276585

Reviewed by Michael Catanzaro.

Amazon Luna (luna.amazon.com) assumes that the user agent is on MacOS if
`window.webkit` is defined, leading to JS calls that only work in the
Apple Mac WebKit port.

`window.webkit` was defined in the GTK MiniBrowser because it registers
a script message handler "aboutData", which makes
`window.webkit.messageHandlers.aboutData.postMessage(msg)` available to
scripts. However, this is only useful for the "about:data" custom page
that is defined in the MiniBrowser.

To have MiniBrowser still working like before but also supporting Amazon
Luna, only register the script message handler when loading "about:*" pages.

* Tools/MiniBrowser/gtk/BrowserTab.c:
(toggleAboutDataScriptMessageHandler):
(loadChanged):
* Tools/MiniBrowser/gtk/main.c:
(activate):

Canonical link: https://commits.webkit.org/281898@main
